File: leetcode/easy/Q155_MinStack.py
class MinStack:
    stack: [int]

    def __init__(self):
        self.stack = []

    def push(self, x: int) -> None:
        self.stack.insert(0, x)

    def pop(self) -> None:
        self.stack.pop(0)

    def top(self) -> int:
        return self.stack[0]

    def getMin(self) -> int:
        min_num = self.stack[0]
        for x in self.stack:
            min_num = min(min_num, x)
        return min_num

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: src/easy/Q155_MinStack.java
package easy;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.Stack;

public class Q155_MinStack {
    class MinStack {
        Stack<Integer> stack;
        Integer minVal = null;
        /** initialize your data structure here. */
        public MinStack() {
            stack = new Stack<>();
        }

        public void push(int val) {
            stack.push(val);
            if(minVal == null) minVal = val;
            minVal = Math.min(minVal, val);
        }

        public void pop() {
            if(minVal.equals(stack.pop())){
                minVal = null;
                Iterator<Integer> iterator = stack.iterator();
                while (iterator.hasNext()){
                    int element = iterator.next();
                    if(minVal == null) minVal = element;
                    minVal = Math.min(minVal, element);
                }
            }
        }

        public int top() {
            return stack.peek();
        }

        public int getMin() {
            return minVal;
        }
    }

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
}
